src/common/java/net/minecraftforge/gradle/common/legacy/LegacyExtension.java

```diff
@@ -0,0 +1,141 @@
+     */
+    public static void runRetrogradleFixes(Project project) {
+        final LegacyExtension config = (LegacyExtension) project.getExtensions().getByName(LegacyExtension.EXTENSION_NAME);
+        // Get the Userdev extension for the run configs
+        final MinecraftExtension minecraft = project.getExtensions().getByType(MinecraftExtension.class);
+        final boolean shouldFixClasspath = config.getFixClasspath().get();
+        final boolean shouldExtractMappings = config.getExtractMappings().get();
+
+        if(shouldFixClasspath) {
+            LOGGER.info("LegacyExtension: Fixing classpath");
+            // create a singleton collection from the jar task's output
+            final FileCollection jar = project.files(project.getTasks().named("jar"));
+
+            minecraft.getRuns().stream()
+                    // get all RunConfig SourceSets
+                    .flatMap(runConfig -> runConfig.getAllSources().stream())
+                    .distinct()
+                    // replace output directories with the jar artifact on each SourceSet's classpath
+                    .forEach(sourceSet -> sourceSet.setRuntimeClasspath(sourceSet.getRuntimeClasspath().minus(sourceSet.getOutput()).plus(jar)));
+        }
+
+        if (shouldExtractMappings) {
+            LOGGER.info("LegacyExtension: Extracting Mappings");
+            // Extracts CSV mapping files to a folder in the build directory for further use
+            // by the <code>net.minecraftforge.gradle.GradleStart.csvDir</code> property.
+            final ExtractZip extractMappingsTask = project.getTasks().create("extractMappings", ExtractZip.class, t -> {
+                t.getZip().fileProvider(project.provider(() -> {
+                    // create maven dependency for mappings
+                    String MAPPING_DEP = "net.minecraft:mappings_{CHANNEL}:{VERSION}@zip";
+                    String coordinates = MAPPING_DEP.replace("{CHANNEL}", minecraft.getMappingChannel().get()).replace("{VERSION}", minecraft.getMappingVersion().get());
+
+                    Dependency dep = project.getDependencies().create(coordinates);
+                    // download and cache the mappings
+                    return project.getConfigurations().detachedConfiguration(dep).getSingleFile();
+                }));
+                // mappings are extracted to <root>/build/<taskName> by default
+                t.getOutput().convention(project.getLayout().getBuildDirectory().dir(t.getName()));
+            });
+            final File csvDir = extractMappingsTask.getOutput().get().getAsFile();
+
+            // attach the csvDir location property to each run configuration
+            minecraft.getRuns().configureEach(run -> run.property("net.minecraftforge.gradle.GradleStart.csvDir", csvDir));
+
+            // execute extractMappings before each run task
+            project.getTasks().getByName("prepareRuns").dependsOn(extractMappingsTask);
```

Comment:
This only runs in afterEvaluate; at that point, all tasks are already present, right? This shouldn't have a performance impact if i understand lazy evaluation correctly.

Reply:
That is wrong. A task may be registered but it does not exist until it is needed by either a task invocation, task dependency or when it is queried.
To quote the [docs](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html):
> To avoid creating and configuring tasks, we say that a task is "registered" but not **created**. When a task is in this state, it is known to the build, it can be configured, and references to it can be passed around, but the task object itself has not actually been created, and none of its configuration actions have been executed.
It will remain in this state until something in the build needs the instantiated task object (for instance if the task is executed on the command line or the task is a dependency of a task executed on the command line). If the task object is never needed, then the task will remain in the registered state, and the cost of creating and configuring the task will be avoided.

Reply:
Okay, I see.

I'll fix that, then.